src/gui/mainwindow.cpp

```diff
@@ -249,9 +244,21 @@ void MainWindow::neovimGuiCloseRequest(int status)
 
 void MainWindow::reconnectNeovim()
 {
+	assert(m_nvim);
+	assert(m_shell);
```

Comment:
Is this safe?

Why not write defensive code and explicitly handle `!m_nvim || !m_shell`?

Reply:
I usually avoid that pattern as it blurs the line between a bug and an error. But I won't go into a philosophical debate about it. :) I'll make the change.

Thank you for reviewing my code.

Reply:
I completely agree philosophically :)

What I'm not sure: will this change trigger a segfault that did not occur before? Seems possible.

Generally, I like to use a reference for non-nullable values and pointers as optional values.

Reply:
> What I'm not sure: will this change trigger a segfault that did not occur before? Seems possible.

Just looking at the related code, `ErrorWidget::reconnectNeovim` is emitted when the retry button is clicked. And right after creating the widget `MainWindow::init` is already called. So, it seems that there's no way that it would and if it does it's certainly a bug.

I think it's safe to leave it as it is, but if you still think I should add the additional check, I will.

Reply:
The `m_nvim` assert is valid I think, since we deference `m_nvim` below.

There is no need for the `m_shell` assert. The existing code is valid, and less strict. A future author could easily make a mistake with the new structure and introduce a case that crashes. Let's avoid this.

`Q_ASSERT` is prefered in Qt code. It provides better error handling, and will be disabled in `Release` mode.

Reply:
> There is no need for the m_shell assert. The existing code is valid, and less strict. A future author could easily make a mistake with the new structure and introduce a case that crashes. Let's avoid this.

I'll make the changes. I just want to comment as a professional curiosity.

Isn't a case that crashes better than a case that hides one? If that mistake is made, and it's hidden the system is basically in an unstable state because there's no telling when the crash is going to exhibit itself. 

Again, I don't want to turn it into a big thing. Just taking this as an opportunity to learn.

Reply:
No big deal, I'm happy to discuss. Discussion is good :)

Yes, that's generally good practice.

Here, you're adding a requirement to the code which did not exist before. The requirement that m_shell/m_nvim is or is not nullptr is not obvious from the API (a non-null ptr should be a reference). The requirement is easy to miss.

What does this change accomplish? The old code is more tolerant of fault cases and does not leak. The new code has cases where a possible state will trigger asset(false). Right?

Reply:
> What does this change accomplish? The old code is more tolerant of fault cases and does not leak. The new code has cases where a possible state will trigger asset(false). Right?

That's right. And it's exactly the reason why it's useful. The requirement that an initialization should only happen if it has not been initialized before is a pretty obvious one. Otherwise, what's stopping me from calling `init` in a loop over and over again because for some reason my code worked if I called it twice but not when I called it once? Assertions make it easier to not write lazy code.

I find that assertions make a code base more robust and less error prone. My life has been much easier when QA or another developer comes to be with an assertion than with a description of a problem. Not that you don't agree with this already...

-----

On another note, I fixed all your comments and pushed the commits. This has been a great experience as a first time contributor, I appreciate it.